h3-quinn/src/lib.rs

```diff
@@ -257,189 +181,95 @@ where
-        )
+    fn poll_data(&mut self, cx: &mut Context<'_>) -> Poll<Result<Option<Self::Buf>, Self::Error>> {
+        self.recv.poll_data(cx)
     }
 
-    fn err_code(&self) -> Option<u64> {
-        match self.0 {
-            quinn::ReadError::ConnectionLost(quinn::ConnectionError::ApplicationClosed(
-                quinn_proto::ApplicationClose { error_code, .. },
-            )) => Some(error_code.into_inner()),
-            quinn::ReadError::Reset(error_code) => Some(error_code.into_inner()),
-            _ => None,
-        }
+    fn stop_sending(&mut self, err_code: u64) {
+        self.recv.stop_sending(err_code)
     }
 }
 
-/// Quinn-backed send stream
+/// Send-only stream
 ///
-/// Implements a [`quic::SendStream`] backed by a [`quinn::SendStream`].
+/// A [`quic::SendStream`] backed by [`quinn::SendStream`].
 pub struct SendStream<B: Buf> {
     stream: quinn::SendStream,
     writing: Option<WriteBuf<B>>,
 }
 
-impl<B> SendStream<B>
-where
-    B: Buf,
-{
-    fn new(stream: quinn::SendStream) -> SendStream<B> {
+impl<B: Buf> SendStream<B> {
+    fn new(stream: quinn::SendStream) -> Self {
         Self {
             stream,
             writing: None,
         }
     }
 }
 
-impl<B> quic::SendStream<B> for SendStream<B>
-where
-    B: Buf,
-{
-    type Error = SendStreamError;
+impl<B: Buf> quic::SendStream<B> for SendStream<B> {
+    type Error = SendError;
 
-    fn poll_ready(&mut self, cx: &mut task::Context<'_>) -> Poll<Result<(), Self::Error>> {
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
         if let Some(ref mut data) = self.writing {
             while data.has_remaining() {
-                match ready!(Pin::new(&mut self.stream).poll_write(cx, data.chunk())) {
-                    Ok(cnt) => data.advance(cnt),
-                    Err(err) => {
-                        // We are forced to use AsyncWrite for now because we cannot store
-                        // the result of a call to:
-                        // quinn::send_stream::write<'a>(&'a mut self, buf: &'a [u8]) -> Write<'a, S>.
-                        //
-                        // This is why we have to unpack the error from io::Error below. This should not
-                        // panic as long as quinn's AsyncWrite impl doesn't change.
-                        return Poll::Ready(Err(SendStreamError::Write(
-                            err.into_inner()
-                                .expect("write stream returned an empty error")
-                                .downcast_ref::<WriteError>()
-                                .expect(
-                                    "write stream returned an error which type is not WriteError",
-                                )
-                                .clone(),
-                        )));
+                let mut write_fut = Box::pin(self.stream.write(data.chunk()));
+                match ready!(write_fut.poll_unpin(cx)) {
```

Comment:
You could use a block here: 
```rust
match ready!({
    let mut write_fut = self.stream.write(data.chunk());
    let write_fut = unsafe { Pin::new_unchecked(&mut write_fut)) };
    write_fut.poll(cx)
})
```

So, you don't need to explicitly drop that future.

Reply:
nice catch!